### ISSUE TITLE
[Comando] Sincronizar los showcase de la web con el bot

### ISSUE THREAD

**Issue by SergioRibera** [RustLangES/cangrebot#27]
## Necesita
- #25 
- #26

**Comment by Phosphorus-M** [comment]
Ojo está buena la idea, puede ser

**Comment by DrewSC13** [comment]
Hola equipo

Me gustaría revisar este issue como continuación del trabajo del endpoint `/showcase` del issue #26.

Ya estuve mirando el código actual de `cangrebot` y veo que el endpoint `/showcase` ya está disponible en la API. También vi que este issue depende de #25 y #26, así que quería confirmar cuál sería el siguiente paso esperado para esta sincronización.

¿La idea es implementar algo más desde el lado del bot, como un comando para sincronizar showcases, o este flujo debería trabajarse principalmente desde el repo de la web consumiendo el endpoint `/showcase`?

Quedo atento para avanzar por donde corresponda.

**Comment by Phosphorus-M** [comment]
No estoy seguro realmente cual es el flujo pensado, podría ser un webhook que sea capturado por el Cangrebot y en cada publicación cree un recurso showcase en el bot y desde la pagina se consuma pegandole al endpoint, o podría ser un comando como dices en lugar de usar un webhook.

Flujo 1:

Re utilizando el flujo actual podríamos hacer lo mismo, actualmente en la pagina se obtiene el contenido de forma estatica, es decir, hay un submodulo de este repositorio:
https://github.com/RustLangES/proyectos-comunitarios

Si se quiere continuar con el flujo actual seria que se genere un PR automatico mediante un dispatch de github action, se puede ver un ejemplo aquí:
https://github.com/nuria-fl/jamstack-cfp

Y que al mismo tiempo dispare otra github action para actualizar el submodulo del repositorio de la pagina:
https://github.com/RustLangES/RustLangES.github.io/tree/main

La ventaja es que todo se resolveria en tiempo de compilación y evitariamos hacer fetch al backend del cangrebot, exponemos menos la API y todo eso.

---

Flujo 2:

Agregar el endpoint con el recurso, el recurso se crea mediante un webhook o un comando, y se hace un fetch en este componente (posiblemente):

https://github.com/RustLangES/RustLangES.github.io/blob/main/src/components/community_projects.rs


---

Flujo 3:

No tengo idea, quizás se les ocurre algo más pro o más simple

Para mi cualquiera de las tres va bien

**Comment by DrewSC13** [comment]
Gracias por el detalle, @Phosphorus-M 

Entonces entiendo que el flujo 1 consistiría en mantener `proyectos-comunitarios` como fuente de datos, consumirlo desde `RustLangES.github.io` vía submódulo y resolver la sincronización durante build/deploy mediante GitHub Actions, en lugar de hacer fetch al backend de Cangrebot en runtime.

Me parece el camino más limpio inicialmente porque mantiene la web estática, evita dependencia directa del bot/API al momento de renderizar y permite automatizar la actualización con un workflow tipo `repository_dispatch` o actualización del submódulo.

Voy a revisar el repo `proyectos-comunitarios`, el submódulo en `RustLangES.github.io` y el componente `community_projects.rs` para entender el flujo actual.

**Comment by DrewSC13** [comment]
Ya revisé el flujo actual:

- `RustLangES.github.io` consume `RustLangES/proyectos-comunitarios` como submódulo en `extras`.
- `build.rs` genera `src/extras/projects.rs` desde los `.toml` del submódulo durante build.
- `sitemap_hook.yml` ya tiene soporte para `repository_dispatch` y ejecuta `git submodule update --remote`.

Como primer cambio pequeño, propongo extender ese workflow para escuchar también un evento `community_projects_changed`, reutilizando el flujo actual de actualización del submódulo + build/deploy.

Luego, en una segunda parte, se podría agregar un workflow en `proyectos-comunitarios` que dispare ese `repository_dispatch` cuando haya cambios en `main`, siempre que exista un secret/token de la organización con permisos para ejecutar workflows en `RustLangES.github.io`.

¿Les parece bien empezar con ese cambio en `RustLangES.github.io`?

**Comment by SergioRibera** [comment]
@DrewSC13 me gusta lo que propones, ya mergee el cambio en la web, luego vemos como sincronizar el foro en el discord, alguna idea?

**Comment by DrewSC13** [comment]
Gracias me parece súper, @SergioRibera 

Creo que una opción bastante limpia sería mantener proyectos-comunitarios como fuente de verdad y hacer que Cangrebot sincronice desde ahí hacia el foro de Discord.

La idea sería algo así:

* Cuando se actualice proyectos-comunitarios, un workflow detecta los cambios.
* Ese workflow toma los proyectos definidos en los .toml.
* Luego envía esa información a Cangrebot usando el endpoint /showcase, o quizá un endpoint más específico para este caso, como POST /showcase/sync.
* Cangrebot se encargaría de crear en el foro de Discord los proyectos que todavía no existan.

Del lado del bot, creo que lo más importante sería que la sincronización sea idempotente para no duplicar publicaciones si el workflow se ejecuta más de una vez. Para eso se podría usar una referencia estable del proyecto, como el link, el nombre normalizado o incluso la ruta del .toml, y con eso verificar si el proyecto ya existe en el foro antes de crearlo.

Creo que ssi se mantiene la web simple y estática, deja el repo de proyectos como punto central de datos y evita tener que mantener manualmente lo mismo en varios lugares. Discord quedaría más como una salida sincronizada, no como otra fuente que haya que actualizar aparte.

Qué les parece este enfoque!? Si ven algo que no sea viable, o si hay alguna restricción que no estoy considerando, cualquier observación es bienvenida para ajustar la propuesta antes de avanzar.

**Comment by SergioRibera** [comment]
@DrewSC13 creo que ya habías hecho una PR antes para agregar el endpoints de showcase en el cangrebot, podemos usar ese para actualizar desde Github hacia Discord, ahora nos queda desde Discord hacia Github (porque aqui es donde se genera la pagina), tenemos que hayar una manera en que no se repita al momento de crear el foro en Discord cuando se invoca desde Github, justamente lo que mencionas, quizás allí sea bueno tener un archivo a modo de cache en donde se guarden los id de los foros en el Discord, no sé, pensando ando

**Comment by DrewSC13** [comment]
Sí, es una gran idea @SergioRibera podríamos manejarlo con un endpoint aparte, algo tipo `POST /showcase/sync`, y dejar el `/showcase` actual para crear un showcase individual. La idea sería que el sync reciba la lista de proyectos desde `proyectos-comunitarios`, y que Cangrebot compare esa información contra un cache simple para saber cuáles ya tienen post en el foro y cuáles faltan. Ese cache podría guardar una referencia estable del proyecto, como el `link`, el nombre normalizado o la ruta del `.toml`, asociada al ID del post/thread en Discord.

Algo tipo: `proyectos/web/graphul.toml -> discord_thread_id`
Con eso evitamos duplicados si el workflow se ejecuta más de una vez, y tampoco haría falta recorrer todo el foro cada vez.

**Comment by SergioRibera** [comment]
@DrewSC13 si, me agrada

**Comment by DrewSC13** [comment]
Perfecto @SergioRibera, entonces avanzo con esa idea. Voy a preparar una primera versión en Cangrebot con un endpoint de sincronización tipo `POST /showcase/sync`.
La idea inicial será mantenerlo simple:
- recibir una lista de proyectos;
- usar una key estable por proyecto (`link`, nombre normalizado o ruta del `.toml`);
- guardar un cache local con la relación `project_key -> discord_thread_id`;
- crear solo los posts que todavía no existan en el foro;
- dejar la actualización de posts existentes para una segunda iteración si hace falta.

**Comment by DrewSC13** [comment]
Ahora que el endpoint de sync ya está listo en Cangrebot, el siguiente paso sería implementar el workflow en RustLangES.github.io para disparar automáticamente la sincronización cuando haya cambios en proyectos-comunitarios.

Con eso ya quedaría el flujo completo end-to-end entre GitHub → Cangrebot → Discord.

**Comment by SergioRibera** [comment]
@DrewSC13 exactamente, con eso ya quedaría el flujo desde Github hacia Discord, para el Workflow de sync desde la web tendría que tener un triggees manual igual y creo que tendría que ser en el repo de https://github.com/RustLangES/proyectos-comunitarios 👀

**Comment by DrewSC13** [comment]
Si súper @SergioRibera entiendo, entonces quedaría bien como siguiente paso, agregar el trigger en proyectos-comunitarios para disparar este repository_dispatch cuando corresponda.

Agregué el workflow en RustLangES.github.io como receptor del evento community_projects_changed: actualiza el submódulo extras, arma el payload desde los .toml de proyectos-comunitarios y llama al endpoint POST /showcase/sync de Cangrebot.

Quedo pendiente a la revisión de ese PR para poder continuar.
